[Bug]: Easy Connect fails to connect if the same player joins from a different machine.

### Describe the Bug

This happens when using the Easy Connect feature for player keys. If a player already exists in the player database, but try to join from a different machine than previously (thus having a different key), the connecting client is simultaneously disconnected and told that the host must verify their pin. On the host side, there is no indication of a join attempt except in the logs.

If using WebRTC, the situation is almost the same except that the client is disconnected with no pin message shown.

This does not occur on 1.15.2.

### To Reproduce

1. Start server using the Easy Connect feature.
2. Connect a second MT instance to the server on some computer (can be the same as the one running the server).
3. Shutdown all MT instances.
4. Start the server again.
5. On a computer other than the one used in (2), connect an MT instance to the server using the same player name as in (2).

### Expected Behaviour

If the same player changes machines, they should still be able to connect to an Easy Connect server if the GM verifies them.

### Screenshots

This is what the connecting player looks like when the error happens on a socket server:
![Image](https://github.com/user-attachments/assets/906b2754-aa56-4cbe-86fa-fcbca83e4271)

WebRTC is the same except only the error dialog show.

### MapTool Info

1.16.0-beta.4

### Desktop

Linux Mint 22.1

### Additional Context

I got some stack traces on the latest develop code. These should line up very closely with the 1.16-beta.4 code.

Server-side for a socket server:
<details>
<pre>
12:29:17.636 (ServerHandshake.java:225) [SocketConnection.ReceiveThread] WARN  - Unexpected exception during server handshake
java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Password.publicKeyNotFound
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073) ~[?:?]
	at net.rptools.maptool.server.ServerHandshake.sendAsymmetricKeyAuthType(ServerHandshake.java:527) ~[main/:?]
	at net.rptools.maptool.server.ServerHandshake.handle(ServerHandshake.java:393) ~[main/:?]
	at net.rptools.maptool.server.ServerHandshake.handleMessage(ServerHandshake.java:206) [main/:?]
	at net.rptools.clientserver.simple.connection.AbstractConnection.dispatchMessage(AbstractConnection.java:109) [clientserver.jar:?]
	at net.rptools.clientserver.simple.connection.AbstractConnection.dispatchCompressedMessage(AbstractConnection.java:115) [clientserver.jar:?]
	at net.rptools.clientserver.simple.connection.SocketConnection$ReceiveThread.run(SocketConnection.java:164) [clientserver.jar:?]
Caused by: java.lang.IllegalArgumentException: Password.publicKeyNotFound
	at net.rptools.maptool.model.player.PasswordFilePlayerDatabase.lambda$getPublicKey$8(PasswordFilePlayerDatabase.java:683) ~[main/:?]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188) ~[?:?]
12:29:17.640 (MapToolServer.java:234) [SocketConnection.ReceiveThread] ERROR - Client closing: bad handshake
java.util.concurrent.CompletionException: net.rptools.maptool.server.Handshake$Failure: Unexpected error encountered during handshake.
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:874) [?:?]
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841) [?:?]
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) [?:?]
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194) [?:?]
	at net.rptools.maptool.server.ServerHandshake.handleMessage(ServerHandshake.java:227) [main/:?]
	at net.rptools.clientserver.simple.connection.AbstractConnection.dispatchMessage(AbstractConnection.java:109) [clientserver.jar:?]
	at net.rptools.clientserver.simple.connection.AbstractConnection.dispatchCompressedMessage(AbstractConnection.java:115) [clientserver.jar:?]
	at net.rptools.clientserver.simple.connection.SocketConnection$ReceiveThread.run(SocketConnection.java:164) [clientserver.jar:?]
Caused by: net.rptools.maptool.server.Handshake$Failure: Unexpected error encountered during handshake.
	... 4 more
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Password.publicKeyNotFound
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073) ~[?:?]
	at net.rptools.maptool.server.ServerHandshake.sendAsymmetricKeyAuthType(ServerHandshake.java:527) ~[main/:?]
	at net.rptools.maptool.server.ServerHandshake.handle(ServerHandshake.java:393) ~[main/:?]
	at net.rptools.maptool.server.ServerHandshake.handleMessage(ServerHandshake.java:206) ~[main/:?]
	... 3 more
Caused by: java.lang.IllegalArgumentException: Password.publicKeyNotFound
	at net.rptools.maptool.model.player.PasswordFilePlayerDatabase.lambda$getPublicKey$8(PasswordFilePlayerDatabase.java:683) ~[main/:?]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188) ~[?:?]
12:29:17.642 (Router.java:61) [SocketConnection.ReceiveThread] ERROR - Failed to remove connection 10.147.17.207-0 because we do not have that connection
</pre>
</details>

And same for WebRTC (required a minor edit to have the trace be dumped alongside the message):
<details>
<pre>
12:21:56.088 (ServerHandshake.java:225) [WebRTCConnection.WebRTCThread_player@kwvantest] WARN  - Unexpected exception during server handshake
java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Password.publicKeyNotFound
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073) ~[?:?]
	at net.rptools.maptool.server.ServerHandshake.sendAsymmetricKeyAuthType(ServerHandshake.java:527) ~[main/:?]
	at net.rptools.maptool.server.ServerHandshake.handle(ServerHandshake.java:393) ~[main/:?]
	at net.rptools.maptool.server.ServerHandshake.handleMessage(ServerHandshake.java:206) [main/:?]
	at net.rptools.clientserver.simple.connection.AbstractConnection.dispatchMessage(AbstractConnection.java:109) [clientserver.jar:?]
	at net.rptools.clientserver.simple.connection.AbstractConnection.dispatchCompressedMessage(AbstractConnection.java:115) [clientserver.jar:?]
	at net.rptools.clientserver.simple.connection.WebRTCConnection$RTCDataChannelObserverImpl.onMessage(WebRTCConnection.java:572) [clientserver.jar:?]
Caused by: java.lang.IllegalArgumentException: Password.publicKeyNotFound
	at net.rptools.maptool.model.player.PasswordFilePlayerDatabase.lambda$getPublicKey$8(PasswordFilePlayerDatabase.java:683) ~[main/:?]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188) ~[?:?]
12:21:56.093 (MapToolServer.java:234) [WebRTCConnection.WebRTCThread_player@kwvantest] ERROR - Client closing: bad handshake
java.util.concurrent.CompletionException: net.rptools.maptool.server.Handshake$Failure: Unexpected error encountered during handshake.
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:874) [?:?]
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841) [?:?]
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) [?:?]
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194) [?:?]
	at net.rptools.maptool.server.ServerHandshake.handleMessage(ServerHandshake.java:227) [main/:?]
	at net.rptools.clientserver.simple.connection.AbstractConnection.dispatchMessage(AbstractConnection.java:109) [clientserver.jar:?]
	at net.rptools.clientserver.simple.connection.AbstractConnection.dispatchCompressedMessage(AbstractConnection.java:115) [clientserver.jar:?]
	at net.rptools.clientserver.simple.connection.WebRTCConnection$RTCDataChannelObserverImpl.onMessage(WebRTCConnection.java:572) [clientserver.jar:?]
Caused by: net.rptools.maptool.server.Handshake$Failure: Unexpected error encountered during handshake.
	... 4 more
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Password.publicKeyNotFound
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073) ~[?:?]
	at net.rptools.maptool.server.ServerHandshake.sendAsymmetricKeyAuthType(ServerHandshake.java:527) ~[main/:?]
	at net.rptools.maptool.server.ServerHandshake.handle(ServerHandshake.java:393) ~[main/:?]
	at net.rptools.maptool.server.ServerHandshake.handleMessage(ServerHandshake.java:206) ~[main/:?]
	... 3 more
Caused by: java.lang.IllegalArgumentException: Password.publicKeyNotFound
	at net.rptools.maptool.model.player.PasswordFilePlayerDatabase.lambda$getPublicKey$8(PasswordFilePlayerDatabase.java:683) ~[main/:?]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188) ~[?:?]
</pre>
</details>

Client-side there does not appear to be any errors generated.

--- a/clientserver/src/main/java/net/rptools/clientserver/simple/server/WebRTCServer.java
+++ b/clientserver/src/main/java/net/rptools/clientserver/simple/server/WebRTCServer.java
@@ -105,7 +105,7 @@ public class WebRTCServer extends AbstractServer {
       @Override
       public void onError(Exception ex) {
         lastError = "WebSocket error: " + ex.toString() + "\n";
-        log.error("S " + lastError);
+        log.error("S " + lastError, ex);
         // onClose will be called after this method
       }
     };
@@ -147,7 +147,7 @@ public class WebRTCServer extends AbstractServer {
     try {
       fireClientConnect(connection);
     } catch (Exception e) {
-      log.error(e);
+      log.error("Unexpected error while handling new data channel", e);
     }
   }
 

--- a/src/main/java/net/rptools/maptool/server/ServerHandshake.java
+++ b/src/main/java/net/rptools/maptool/server/ServerHandshake.java
@@ -523,19 +523,19 @@ public class ServerHandshake implements Handshake<Player>, MessageHandler {
       } else {
         sendErrorResponseAndNotify(HandshakeResponseCodeMsg.INVALID_PUBLIC_KEY);
       }
-      return;
-    }
-    CipherUtil.Key publicKey = playerDatabase.getPublicKey(player, playerPublicKeyMD5).get();
-    String password = new PasswordGenerator().getPassword();
-    handshakeChallenges[0] =
-        HandshakeChallenge.createAsymmetricChallenge(player.getName(), password, publicKey);
+    } else {
+      CipherUtil.Key publicKey = playerDatabase.getPublicKey(player, playerPublicKeyMD5).get();
+      String password = new PasswordGenerator().getPassword();
+      handshakeChallenges[0] =
+          HandshakeChallenge.createAsymmetricChallenge(player.getName(), password, publicKey);
 
-    var authTypeMsg =
-        UseAuthTypeMsg.newBuilder()
-            .setAuthType(AuthTypeEnum.ASYMMETRIC_KEY)
-            .addChallenge(ByteString.copyFrom(handshakeChallenges[0].getChallenge()));
-    var handshakeMsg = HandshakeMsg.newBuilder().setUseAuthTypeMsg(authTypeMsg).build();
-    sendMessage(State.AwaitingClientPublicKeyAuth, handshakeMsg);
+      var authTypeMsg =
+          UseAuthTypeMsg.newBuilder()
+              .setAuthType(AuthTypeEnum.ASYMMETRIC_KEY)
+              .addChallenge(ByteString.copyFrom(handshakeChallenges[0].getChallenge()));
+      var handshakeMsg = HandshakeMsg.newBuilder().setUseAuthTypeMsg(authTypeMsg).build();
+      sendMessage(State.AwaitingClientPublicKeyAuth, handshakeMsg);
+    }
   }
 
   @Override

--- a/src/main/java/net/rptools/maptool/server/ServerHandshake.java
+++ b/src/main/java/net/rptools/maptool/server/ServerHandshake.java
@@ -523,6 +523,7 @@ public class ServerHandshake implements Handshake<Player>, MessageHandler {
       } else {
         sendErrorResponseAndNotify(HandshakeResponseCodeMsg.INVALID_PUBLIC_KEY);
       }
+      return;
     }
     CipherUtil.Key publicKey = playerDatabase.getPublicKey(player, playerPublicKeyMD5).get();
     String password = new PasswordGenerator().getPassword();

